### PR TITLE
Expose regionName setting in Kinesis receiver configuration

### PR DESCRIPTION
**Pull request by kopiczko**
Hi,

I'd like to have that setting exposed to be able to set DynamoDB table (for KCL) in the same region as the Kinesis stream.

### CODE REVIEW

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala [hunk 1]
```diff
@@ -81,17 +84,19 @@ object KinesisUtils {
...
    *
-   * @return JavaReceiverInputDStream[Array[Byte]]
+   * @return JavaReceiverInputDStream[ Array[Byte] ]
    */
   @Experimental
   def createStream(
-      jssc: JavaStreamingContext, 
-      streamName: String, 
-      endpointUrl: String, 
+      jssc: JavaStreamingContext,
```
Comment (srowen) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: Isn't this changing the API?
I believe you should make a JIRA along with this PR. It's not trivial. 
https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark

**Comment by cfregly** [comment]
@srowen @kopiczko 

This is part of a larger effort to overhaul Kinesis-based streaming slated for 1.4.  Lots of API changes including region, AWS credentials, and application name - as well as upgrading both the AWS Java SDK and the KCL.

Here's the parent jira:  https://issues.apache.org/jira/browse/SPARK-6599.

Here's the related jira that covers the region portion:  https://issues.apache.org/jira/browse/SPARK-6514.

We should definitely try to be backward-compatible even though the API is Experimental.

**Comment by kopiczko** [comment]
Thanks guys for your response. @cfregly I've already answered your comment on Jira. I guess we should move our discussion there. I can improve this implementation to meet requirements in SPARK-6514 and rename PR according to guidelines in wiki.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 2]
```diff
@@ -36,18 +36,19 @@ import com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker
...
  *   to run within a Spark Executor.
  *
  * @param appName  Kinesis application name. Kinesis Apps are mapped to Kinesis Streams
  *                 by the Kinesis Client Library.  If you change the App name or Stream name,
- *                 the KCL will throw errors.  This usually requires deleting the backing  
+ *                 the KCL will throw errors.  This usually requires deleting the backing
  *                 DynamoDB table with the same name this Kinesis application.
  * @param streamName   Kinesis stream name
  * @param endpointUrl  Url of Kinesis service (e.g., https://kinesis.us-east-1.amazonaws.com)
+ * @param regionName   Region name to indicate the location of the Amazon Kinesis service
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: might want to add a note similar to the KCL README.md when describing this param:

"The Amazon DynamoDB table and Amazon CloudWatch metrics associated with your application will also use this region setting."

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala [hunk 3]
```diff
@@ -39,6 +39,7 @@ object KinesisUtils {
    * @param ssc    StreamingContext object
    * @param streamName   Kinesis stream name
    * @param endpointUrl  Url of Kinesis service (e.g., https://kinesis.us-east-1.amazonaws.com)
+   * @param regionName   Region name to indicate the location of the Amazon Kinesis service
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: same here...

"The Amazon DynamoDB table and Amazon CloudWatch metrics associated with your application will also use this region setting."

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala [hunk 4]
```diff
@@ -70,6 +72,7 @@ object KinesisUtils {
    * @param jssc Java StreamingContext object
    * @param streamName   Kinesis stream name
    * @param endpointUrl  Url of Kinesis service (e.g., https://kinesis.us-east-1.amazonaws.com)
+   * @param regionName   Region name to indicate the location of the Amazon Kinesis service
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: and here?  :)

"The Amazon DynamoDB table and Amazon CloudWatch metrics associated with your application will also use this region setting."

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala (hunk 3 above)
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: we may want to consider making a default for this to maintain backward compatibility.

the problem is that the Scala and Java createStream() methods in this helper class will conflict if you use defaults for this.  i had the same issue with initialPositionInStream as well as storageLevel which is why they're not defaults.  

not sure we can do much about it without changing the names of the methods to createScalaStream() and createJavaStream() or equivalent.

**Comment by tdas** [comment]
Any updates on this patch? If you are not able to work on it, mind closing it?

**Comment by tdas** [comment]
This is not needed any more as Spark 1.4.0 has fixed this issue. Mind closing this PR?

**Comment by kopiczko** [comment]
Sorry guys, I currently have no time to work on it. @tdas: Would you mind give a reference how this is solved?

**Comment by kopiczko** [comment]
Sorry guys, I currently have no time to work on it. @tdas: Would you mind giving a reference how this is solved?

**Comment by tdas** [comment]
Sure, take a look at the new KinesisUtils API. 

https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.streaming.kinesis.KinesisUtils$

**Pull request closed without merging**